Feature/hollow producer restore

davesu opened the pull request:
Enhancement to enable HollowProducer to restore blob even after it has existing non empty state 

- Restore after it was already published states
- Add ability to indicate whether Restore Failure is Fatal (throw RuntimeException)
- Add support for announcement metadata
-

Commit:
Enhance Hollow Producer to properly restore at any state

- Reset HollowWriteStateEngine/ObjectMapper (new instances) prior to
restore because restoring on non-empty state is not supported

- Introduce ability to toggle restore failure as fatal

- Add hook to enable metadata as part of announcement

- expose ability to signal force double snapshot on next update

Commit:
HollowProducer Unit Tests

Commit:
Merge remote-tracking branch 'origin/master' into
feature/hollow_producer_restore

# Conflicts:
#	hollow/src/main/java/com/netflix/hollow/api/producer/HollowProducer.java

dkoszewnik commented on hollow/src/main/java/com/netflix/hollow/api/producer/HollowProducer.java:
```diff
@@ -561,7 +625,7 @@ public void cleanup() {
 
     public static interface Announcer {
-        public void announce(long stateVersion);
+        public void announce(long stateVersion, Map<String, String> metadata);
```
Adding metadata tags to announcement feels like unnecessary cognitive load -- we should avoid this. The concept of announcement should remain simple: http://hollow.how/producer-consumer/#announcing-the-state. Specific implementations are free to introduce additional complexity as necessary, but that complexity shouldn't leak into the core producer API.

davesu replied:
This was implemented for the use case to notify client about schema changes on delta.   Any suggestions on an alternative way to propagate this information?

dkoszewnik commented on hollow/src/main/java/com/netflix/hollow/api/producer/HollowProducer.java:
```diff
@@ -358,6 +380,48 @@ private void announce(HollowConsumer.ReadState readState) {
         }
     }
 
+    private class WriteEngineWrapper {
```
I'd like to understand more about why the WriteEngineWrapper is necessary -- if we can do without it and if not, whether there is a better name to express its purpose

davesu replied:
I agree a better name for this class would be good.  

The purpose was to encapsulate classes that needs to be changed on a Restore.  Originally I was going to use WriteState but that is already used.   

Let's try to come up with something better.

dkoszewnik commented on hollow/src/main/java/com/netflix/hollow/api/producer/HollowProducer.java:
```diff
@@ -83,27 +84,34 @@ public HollowProducer(
         listeners.fireProducerInit(currentTimeMillis() - start);
     }
 
+    public void setRestoreFailureFatal(boolean isFatal) {
```
I think that the behavior of the system when restore fails should be defined by the caller to restore, rather than have to configure it here.

davesu replied:
Ok, let me try that.

Commit:
Incorporated review comments

- Renamed WriteEngineWrapper to ProducerDataState

- Removed configuration for restore failure as fatal and make
corresponding API more explicit

Commit:
Cleanup

- Removed ProducerDataState and instead just use ObjectMapper to
encapsulate DataState that needs to be reset at every restore

- Reverted changes to Announcer to support metadata until later point
when there are more use cases

Commit:
Code cleanup - removed unnecessary thingies

Commit:
more cleanup - remove unused imports
